types/src/client.rs

```diff
@@ -0,0 +1,91 @@
+use crate::error::Error;
+use crate::jsonrpc::{self, DeserializeOwned, JsonValue, Params, SubscriptionId};
+use core::marker::PhantomData;
+use futures::channel::{mpsc, oneshot};
+use futures::prelude::*;
+
+/// Active subscription on a Client.
+pub struct Subscription<Notif> {
+	/// Channel to send requests to the background task.
+	pub to_back: mpsc::Sender<FrontToBack>,
+	/// Channel from which we receive notifications from the server, as undecoded `JsonValue`s.
+	pub notifs_rx: mpsc::Receiver<JsonValue>,
+	/// Subscription ID,
+	pub id: SubscriptionId,
+	/// Marker in order to pin the `Notif` parameter.
+	pub marker: PhantomData<Notif>,
+}
+
+/// Message that the Client can send to the background task.
+pub enum FrontToBack {
+	/// Send a one-shot notification to the server. The server doesn't give back any feedback.
+	Notification {
+		/// Method for the notification.
+		method: String,
+		/// Parameters to send to the server.
+		params: Params,
+	},
+
+	/// Send a request to the server.
+	StartRequest {
+		/// Method for the request.
+		method: String,
+		/// Parameters of the request.
+		params: Params,
+		/// One-shot channel where to send back the outcome of that request.
+		send_back: oneshot::Sender<Result<JsonValue, Error>>,
+	},
+
+	/// Send a subscription request to the server.
+	Subscribe {
+		/// Method for the subscription request.
+		subscribe_method: String,
+		/// Parameters to send for the subscription.
+		params: Params,
+		/// Method to use to later unsubscription. Used if the channel unexpectedly closes.
+		unsubscribe_method: String,
+		/// When we get a response from the server about that subscription, we send the result on
+		/// this channel. If the subscription succeeds, we return a `Receiver` that will receive
+		/// notifications.
+		send_back: oneshot::Sender<Result<(mpsc::Receiver<JsonValue>, SubscriptionId), Error>>,
+	},
+
+	/// When a subscription channel is closed, we send this message to the background
+	/// task to mark it ready for garbage collection.
+	// NOTE: It is not possible to cancel pending subscriptions or pending requests.
+	// Such operations will be blocked until a response is received or the background
+	// thread has been terminated.
```

Comment:
Is this a possible DoS attack vector? I guess not, a malicious user can saturate their ws connection with subscription requests but not impact other users' connection right?

Reply:
yeah, it will send a message to background thread which in turn makes a request to the server but it's up to the server to handle the request and connection. If the connection gets closed by the server the user has to restart the client to get it working again.

So nothing really related to this, from a malicious users point-of-view it is better to make requests to DoS a server but it might be possible to make the client less responsive/hung by continuously spawn new subscriptions and close them directly.